### PR TITLE
Speed up TimeDistributedDense, 3D Softmax and RNNs

### PR DESCRIPTION
**Pull request by farizrahman4u**
Found some serious bottlenecks in core parts of Keras:

* Currently `TimeDistributedDense` is implemented using `K.rnn()`, i.e, each vector in the sequence is dot with `W` sequentially, one after the other. Since each time step of the rnn is independent of the previous time step (no state involved), instead of looping through the sequence, it is possible to obtain the output in one single huge matrix operation.

* Similarly, softmax activiation for 3D data is also implemented using `rnn()`, which translates to 1 softmax op per vector.

* Now if you were to combine these two, i.e, `TimeDistributedDense(10, activation='softmax')`, Keras will loop through each of your sequence **twice**, one for the dot operation and the other for the activation. Sad ):

* Instead of making use of shaping ops such as `permute_dimesions` (which come at no cost), Dot and sum ops are used to initialize RNN hidden states.

### CODE REVIEW

File: keras/layers/core.py [hunk 1]
```diff
@@ -1121,16 +1121,19 @@ def output_shape(self):
...
+            Y = z + self.b
+            Y = self.activation(Y)
+            return Y
+        elif K._BACKEND == 'tensorflow':
+            shape = self.input_shape
+            X = K.reshape(X, (-1, shape[-1]))
+            Y = K.dot(X, self.W) + self.b
+            Y = K.reshape(Y, (-1, shape[1], shape[2]))
+            Y = self.activation(Y)
+            return Y
```
Comment (fchollet) [inline keras/layers/core.py]: It would be bad practice to have segments of code dedicated to one backend or another. Would there be a performance impact to just using the `K`-based code? 

Or could we reimplement `tensordot` in TensorFlow? (and make it part of the backend).

Comment (farizrahman4u) [inline keras/layers/core.py]: Its not about performance, the K based code uses input length. But when using theano backend it is possible not to provide the time dimension.

**Comment by farizrahman4u** [comment]
@fchollet
> It would be bad practice to have segments of code dedicated to one backend or another. Would there be a performance impact to just using the K-based code?

Its not about performance. The K based code needs the input length, but when using theano backend it is possible not to provide the time dimension.

**Comment by farizrahman4u** [comment]
@fchollet it's all K code now. Happy ? :)

**Comment by farizrahman4u** [comment]
@fchollet Done. It's all K code now. Works even if `input_length` is `None`. Happy? :)

**Comment by farizrahman4u** [comment]
Having a tough time fixing tensorflow issues. Slicing isn't as easy as theano. `X[0][0]` and `X[0, 0]` do not work. Can not reshape if there are unknown dims in the reshape tensor. Crappy. To make things worse, i am on Windows and do not have TF installed, so have to rely on Travis to test my code :(

**Comment by farizrahman4u** [comment]
@fchollet  Finally, the tests are passing.

File: keras/layers/recurrent.py [hunk 2]
```diff
@@ -120,11 +120,9 @@ def get_constants(self, X, train=False):
         return None
 
     def get_initial_states(self, X):
-        # build an all-zero tensor of shape (samples, output_dim)
```
Comment (fchollet) [inline keras/layers/recurrent.py]: This comment is still necessary.

File: keras/layers/core.py [hunk 3]
```diff
@@ -1121,16 +1121,18 @@ def output_shape(self):
...
+        x = K.reshape(X, (-1, self.input_shape[-1]))
+        Y = K.dot(x, self.W) + self.b
+        input_length = self.input_shape[1]
+        if input_length:
+            Y = K.reshape(Y, (-1, input_length, self.output_shape[-1]))
+        else:
+            z = X[:, :, 0]
+            z = K.pack([z] * self.output_shape[-1])
+            z = K.permute_dimensions(z, (1, 2, 0))
+            Y = K.reshape(Y, K.shape(z))
```
Comment (fchollet) [inline keras/layers/core.py]: This part of the code isn't very readable and so would benefit from explanatory comments (similar to how there was one comment per line in `get_initial_states` from `Recurrent`).

**Comment by fchollet** [comment]
Looks good, although readability of `TimeDistributed` is taking a hit. Do you have any numbers on performance before/after with a test network?

**Comment by farizrahman4u** [comment]
@fchollet Numbers won't be necessary, we just reduced time complexity from O(n^4) to O(n^3). Its massive. (Consider stacked `TimeDistributedDense` layers). Working on the comments.

**Comment by farizrahman4u** [comment]
Done.

**Comment by fchollet** [comment]
Thanks. Looks good to me, please squash your commits.

**Comment by EderSantana** [comment]
:+1:  @farizrahman4u

**Comment by farizrahman4u** [comment]
I messed up while squashing. My branch looks fine.. [pztch-25](https://github.com/farizrahman4u/keras/tree/patch-25), but something happened to the pull request. Anyway moved to #1812.

File: keras/activations.py [hunk 4]
```diff
@@ -7,13 +7,9 @@ def softmax(x):
...
         return K.softmax(x)
     elif ndim == 3:
-        # apply softmax to each timestep
-        def step(x, states):
-            return K.softmax(x), []
-        last_output, outputs, states = K.rnn(step, x,
-                                             [],
-                                             mask=None)
-        return outputs
+        e = K.exp(x)
```
Comment (jruales) [inline keras/activations.py]: I believe this is prone to overflowing. It should be something like `e = K.exp(x - K.max(x, axis=-1, keepdims=True))`.

Comment (farizrahman4u) [inline keras/activations.py]: Right. Thats equivalent to multiplying numerator and denominator of softmax expression by `e ^ (-max(x))`, so output will be unchanged.

**Comment by farizrahman4u** [comment]
Travis not working

**Comment by fchollet** [comment]
It happens. For now, just squash and rebase, we'll re-run the tests when we can.

**Comment by farizrahman4u** [comment]
Squashed to 3 commits.

**Comment by fchollet** [comment]
Cool, merged.
